Add properties to access Client attributes

PierreF opened the pull request:
Fix #764, add properties for
* host
* port
* keepalive
* transport
* protocol
* connect_timeout
* username
* password
* max_inflight_messages
* max_queued_messages
* will_topic
* will_payload
* logger

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
 
+    @property
+    def host(self) -> str:
+        "Host we try to connect to. If don't yet called connect() return empty string"
```
```suggestion
        """Host to connect to. If `connect()` hasn't been called yet, returns an empty string."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def port(self) -> int:
+        "TCP port we try to connect to."
```
```suggestion
        """Broker TCP port to connect to."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+    @port.setter
+    def port(self, value: int) -> None:
+        """
+        Update post. This will only be used on future (re)connection. You should probably
```
```suggestion
        Update port. This will only be used on future (re)connection. You should probably
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def keepalive(self) -> int:
+        "Keepalive in seconds used by the client "
```
```suggestion
        """Client keepalive interval (in seconds)."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+    def keepalive(self, value: int) -> None:
+        "Update keepalive. It's behavior is undefined if the connection is already open"
+        if self._sock is not None:
+            # The issue here is that the previous value of keepalive matter to possibly
+            # sent ping packet.
+            warnings.warn(
+                "updating keepalive on established connection is not supported",
+                stacklevel=2,
+            )
+            self._sock.settimeout(value)
```
I think this should just raise a RuntimeError if the behavior is undefined?

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @keepalive.setter
+    def keepalive(self, value: int) -> None:
+        "Update keepalive. It's behavior is undefined if the connection is already open"
```
```suggestion
        """Update the client keepalive interval. This may not be called if the connection is already open."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def transport(self) -> str:
+        'Transport used for the connection, could be "tcp" or  "websockets".'
```
I'd rather skip the "could be" since we don't really mandate it.
```suggestion
        """Transport method used for the connection."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def protocol(self) -> int:
+        "Protocol version used (MQTT v3, MQTT v3.11, MQTTv5)"
```
This doesn't specify what the returned values will be, but I suspect this PR will only be merged after #791, which adds an enumeration for the protocol version?
```suggestion
        """Protocol version used (MQTT v3, MQTT v3.11, MQTTv5)"""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def connect_timeout(self) -> float:
+        "Timeout used to establish the TCP (& TLS / websocket if enabled) in seconds"
```
```suggestion
        """Connection establishment timeout in seconds"""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def username(self) -> str | None:
+        "Return the username use to connect to MQTT broken or None if not credenials are used."
```
```suggestion
        """The username used to connect to the MQTT broker, or None if no username is used."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def password(self) -> str | None:
+        "Return the password use to connect to MQTT broken or None if not password are used."
```
```suggestion
        """The password used to connect to the MQTT broker, or None if no password is used."""
```

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def max_inflight_messages(self) -> int:
+        "Maximum number of message with QoS > 0 that can be part way through their network flow at once"
```
```suggestion
        """Maximum number of messages with QoS > 0 that can be partway through the network flow at once"""
```
... but what does "partway through the network flow" mean? It's not quite clear.

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+
+    @property
+    def max_queued_messages(self) -> int:
+        "Maximum number of message with QoS > 0 that can be part way through their network flow at once"
```
```suggestion
        """Maximum number of messages with QoS > 0 that can be partway through the network flow at once"""
```
... same concern about "partway through the network flow" – and then, this string is exactly the same as `max_inflight_messages`, so I'm not sure what the actual difference is? I should hope there is a difference.

PierreF replied:
Here it's a copy/paste mistake I made. Should the same description of max_queued_messages_set which don't imply network, only internal queue.

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+    @max_queued_messages.setter
+    def max_queued_messages(self, value: int) -> None:
+        "Update max_queued_messages. It's behavior is undefined if the connection is already open"
+        if self._sock is not None:
+            # Not tested.
+            warnings.warn(
+                "updating max_queued_messages on established connection is not supported",
+                stacklevel=2,
+            )
```
Should just raise a RuntimeError IMO.

akx commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+    @max_inflight_messages.setter
+    def max_inflight_messages(self, value: int) -> None:
+        "Update max_inflight_messages. It's behavior is undefined if the connection is already open"
+        if self._sock is not None:
+            # Not tested. Some doubt that everything is okay when max_inflight change between 0
+            # and > 0 value because _update_inflight is skipped when _max_inflight_messages == 0
+            warnings.warn(
+                "updating max_inflight_messages on established connection is not supported",
+                stacklevel=2,
+            )
```
Should just raise a RuntimeError.

akx reviewed (changes requested):
Grammatical etc. fixes within – and also, I'm not sure there should be setters at all for the properties which won't take an effect until reconnect? Are there actual use cases for those?

PierreF commented:
I've added setter, because some properties can't be changed without setter (at least connection_timeout), and then I've added setter for other for consistency. But I'm also unsure this will be very useful.

PierreF commented on src/paho/mqtt/client.py:
```diff
@@ -625,6 +624,196 @@ def __init__(self, client_id="", clean_session=None, userdata=None,
+        if self._will_topic is None:
+            return None
+
+        return self._will_topic.decode("utf-8")
```
It's also odd that we use `self._will_TOPIC` in the will_PAYLOAD property. Copy/paste is not always your friend :)

PierreF commented:
I'm going with disallowing (raise RuntimeError) changing properties that need to reconnect. It's only allowed to change them when not connected or *connecting*. It still disallowed if the connection is currently down because a loop_forever or loop_start might reconnect concurrently. It's required to disconnect() which ensure no automatic reconnection occur
